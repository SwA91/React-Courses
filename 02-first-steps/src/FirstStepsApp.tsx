import { ItemCounter } from "./shopping-cart/ItemCounter";

interface ItemInCart {
  productName: string;
  quantity: number;
}
const itemsInCart: ItemInCart[] = [
  {
    productName: " nintendo",
    quantity: 1,
  },
  {
    productName: " playstation 4",
    quantity: 2,
  },
  {
    productName: " xbox 260",
    quantity: 3,
  },
];
export function FirstStepsApp() {
  return (
    <>
      <h1>carrito compras</h1>
      {itemsInCart.map(({ productName, quantity }) => (
        <ItemCounter
          key={productName}
          name={productName}
          quantity={quantity}
        ></ItemCounter>
      ))}
      {/* <ItemCounter name="nintento"></ItemCounter>
      <ItemCounter name="ps5" quantity={3}></ItemCounter>
      <ItemCounter name="xbox" quantity={2}></ItemCounter> */}
    </>
  );
}
