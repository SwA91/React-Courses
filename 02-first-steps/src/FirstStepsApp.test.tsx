import { render, screen } from "@testing-library/react";
import { afterEach, describe, expect, test, vi } from "vitest";
import { FirstStepsApp } from "./FirstStepsApp";

// simula retornos ej. http
const mockItemCounter = vi.fn((props: unknown) => {
  return (
    <div
      data-testid="item-counters"
      name={props.name}
      quantity={props.quantity}
    ></div>
  );
});

// mockeamos la importaciÃ³n del componente en el componente padre
vi.mock("./shopping-cart/ItemCounter", () => ({
  ItemCounter: (props: unknown) => mockItemCounter(props),
}));

// componente ficticio
// alteramos las importaciones del render que hace por debajo
// vi.mock("./shopping-cart/ItemCounter", () => ({
// capturo y regreso lo que quiera
//   ItemCounter: (props: unknown) => (
// <div
//   data-testid="item-counters"
//   name={props.name}
//   quantity={props.quantity}
// ></div>
//   ),
// }));

describe("FirstStepApp", () => {
  afterEach(() => {
    // limpiar los mocks despues de cada test
    vi.clearAllMocks();
  });

  test("should match snapshot", () => {
    const { container } = render(<FirstStepsApp></FirstStepsApp>);
    expect(container).toMatchSnapshot();
  });

  test("should render the correct number of itemcounter components", () => {
    render(<FirstStepsApp></FirstStepsApp>);
    const itemCounters = screen.getAllByTestId("item-counters");
    screen.debug();

    expect(itemCounters.length).toBe(3);
  });

  test("should render ItemCounter with correct props", () => {
    render(<FirstStepsApp></FirstStepsApp>);

    // controlamos cuantas veces se llamo al mock
    expect(mockItemCounter).toHaveBeenCalledTimes(3);
    expect(mockItemCounter).toHaveBeenCalledWith({
      name: " nintendo",
      quantity: 1,
    });
    expect(mockItemCounter).toHaveBeenCalledWith({
      name: " playstation 4",
      quantity: 2,
    });
    expect(mockItemCounter).toHaveBeenCalledWith({
      name: " xbox 260",
      quantity: 3,
    });
  });
});
