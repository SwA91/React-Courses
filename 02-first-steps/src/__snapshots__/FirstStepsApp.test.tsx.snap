// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FirstStepApp > should match snapshot 1`] = `
<div>
  <h1>
    carrito compras
  </h1>
  <div
    data-testid="item-counters"
    name=" nintendo"
    quantity="1"
  />
  <div
    data-testid="item-counters"
    name=" playstation 4"
    quantity="2"
  />
  <div
    data-testid="item-counters"
    name=" xbox 260"
    quantity="3"
  />
</div>
`;
